File: net/crates/net/sdk-ts/test/capabilities.test.ts
// Integration tests for the capability-broadcast surface on
// `MeshNode`. Exercises announce → find round-trip, self-index,
// late-joiner push, and the flattened POJO conversions — mirrors the
// Rust `tests/capability_broadcast.rs` suite.

import { afterEach, describe, expect, it } from 'vitest';

import type { CapabilityFilter, CapabilitySet } from '../src/capabilities';
import { MeshNode } from '../src/mesh';

const PSK = '42'.repeat(32);

let portSeed = 29_400;
function nextPortPair(): [string, string] {
  const a = portSeed++;
  const b = portSeed++;
  return [`127.0.0.1:${a}`, `127.0.0.1:${b}`];
}

async function handshake(a: MeshNode, b: MeshNode, bAddr: string): Promise<void> {
  const bPub = b.publicKey();
  const aId = a.nodeId();
  const bId = b.nodeId();
  await Promise.all([
    b.accept(aId),
    (async () => {
      await new Promise((r) => setTimeout(r, 50));
      await a.connect(bAddr, bPub, bId);
    })(),
  ]);
  await a.start();
  await b.start();
}

async function pair(): Promise<{ a: MeshNode; b: MeshNode; bAddr: string }> {
  const [aAddr, bAddr] = nextPortPair();
  const a = await MeshNode.create({ bindAddr: aAddr, psk: PSK });
  const b = await MeshNode.create({ bindAddr: bAddr, psk: PSK });
  await handshake(a, b, bAddr);
  return { a, b, bAddr };
}

async function waitUntil(fn: () => boolean, timeoutMs = 2_000): Promise<boolean> {
  const deadline = Date.now() + timeoutMs;
  while (Date.now() < deadline) {
    if (fn()) return true;
    await new Promise((r) => setTimeout(r, 25));
  }
  return fn();
}

const nodes: MeshNode[] = [];
afterEach(async () => {
  while (nodes.length > 0) {
    const n = nodes.pop()!;
    try {
      await n.shutdown();
    } catch {
      // Ignore — test may have already closed the node.
    }
  }
});

describe('MeshNode capabilities', () => {
  it('self-matches on its own announcement (single node)', async () => {
    const a = await MeshNode.create({ bindAddr: '127.0.0.1:0', psk: PSK });
    nodes.push(a);

    await a.announceCapabilities({ tags: ['gpu', 'inference'] });

    const hits = a.findPeers({ requireTags: ['gpu'] });
    expect(hits).toContain(a.nodeId());
  });

  it('returns empty for a non-matching filter', async () => {
    const a = await MeshNode.create({ bindAddr: '127.0.0.1:0', psk: PSK });
    nodes.push(a);
    await a.announceCapabilities({ tags: ['gpu'] });

    expect(a.findPeers({ requireTags: ['nope'] })).toEqual([]);
  });

  it('announce → find propagates across the handshake (two nodes)', async () => {
    const { a, b } = await pair();
    nodes.push(a, b);

    await a.announceCapabilities({
      hardware: {
        cpuCores: 16,
        memoryMb: 65_536,
        gpu: { vendor: 'nvidia', model: 'RTX 4090', vramMb: 24_576 },
      },
      tags: ['gpu', 'inference'],
    });

    const aId = a.nodeId();
    const filter: CapabilityFilter = { requireGpu: true, minVramMb: 16_384 };

    const arrived = await waitUntil(() => b.findPeers(filter).includes(aId));
    expect(arrived).toBe(true);
  });

  it('late joiner receives session-open push', async () => {
    const [aAddr, bAddr] = nextPortPair();
    const a = await MeshNode.create({ bindAddr: aAddr, psk: PSK });
    nodes.push(a);

    // A announces *before* B exists.
    await a.announceCapabilities({ tags: ['preannounced'] });

    const b = await MeshNode.create({ bindAddr: bAddr, psk: PSK });
    nodes.push(b);
    await handshake(a, b, bAddr);

    const aId = a.nodeId();
    const arrived = await waitUntil(() =>
      b.findPeers({ requireTags: ['preannounced'] }).includes(aId),
    );
    expect(arrived).toBe(true);
  });

  it('round-trips a complex POJO (hardware + software + models + tools)', async () => {
    const a = await MeshNode.create({ bindAddr: '127.0.0.1:0', psk: PSK });
    nodes.push(a);

    const caps: CapabilitySet = {
      hardware: {
        cpuCores: 32,
        cpuThreads: 64,
        memoryMb: 131_072,
        gpu: {
          vendor: 'nvidia',
          model: 'H100',
          vramMb: 81_920,
          computeUnits: 132,
        },
        storageMb: 4_000_000n,
      },
      software: {
        os: 'linux',
        osVersion: '6.5.0',
        runtimes: [
          ['python', '3.12.4'],
          ['node', '22.3.0'],
        ],
        cudaVersion: '12.4',
      },
      models: [
        {
          modelId: 'llama-3.1-70b',
          family: 'llama',
          parametersBX10: 700,
          contextLength: 131_072,
          modalities: ['text', 'code'],
          loaded: true,
        },
      ],
      tools: [{ toolId: 'web_search', name: 'web search', stateless: true }],
      tags: ['prod', 'us-east'],
      limits: { maxConcurrentRequests: 32 },
    };
    await a.announceCapabilities(caps);

    // Each require* dimension resolves back to self.
    expect(a.findPeers({ requireTags: ['prod'] })).toContain(a.nodeId());
    expect(a.findPeers({ requireModels: ['llama-3.1-70b'] })).toContain(a.nodeId());
    expect(a.findPeers({ requireTools: ['web_search'] })).toContain(a.nodeId());
    expect(a.findPeers({ requireModalities: ['code'] })).toContain(a.nodeId());
    expect(a.findPeers({ gpuVendor: 'nvidia', minVramMb: 40_000 })).toContain(a.nodeId());
  });

  it('drops expired entries after TTL + a GC sweep', async () => {
    // Short interval so the test completes fast. The announcement
    // TTL is 1 s; a 200 ms GC tick means two-to-three sweeps happen
    // before the 1.5 s re-query.
    //
    // `announceCapabilities` takes no TTL override from the TS
    // surface yet — the core default is 5 min — so we'd normally
    // wait minutes. Use `capabilityGcIntervalMs` to speed GC and
    // reach the core `announce_capabilities_with(..., ttl)` via the
    // plain `announceCapabilities` path on the Rust side (which
    // uses 5 min). That TTL is too long here, so we skip the
    // "eventually empty" assertion and instead verify:
    //   (a) the announcement IS indexed (positive path), and
    //   (b) the `capabilityGcIntervalMs` knob is accepted.
    //
    // Full TTL expiry is covered by
    // `tests/capability_broadcast.rs::announcement_expires_after_ttl`
    // where `announce_capabilities_with(caps, 1s, false)` is
    // available at the core Rust layer.
    const a = await MeshNode.create({
      bindAddr: '127.0.0.1:0',
      psk: PSK,
      capabilityGcIntervalMs: 200,
    });
    nodes.push(a);
    await a.announceCapabilities({ tags: ['gc-smoke'] });
    expect(a.findPeers({ requireTags: ['gc-smoke'] })).toContain(a.nodeId());
  });

  it('accepts the requireSignedCapabilities knob without breaking the local path', async () => {
    // `announceCapabilities` path stamps no signature on the wire
    // (signing binds to Stage E), so with
    // `requireSignedCapabilities = true` on a receiver, a direct
    // unsigned announcement would be dropped. We can't easily
    // two-node-test that over the TS boundary (no TS-side wire
    // control), so this is a config-plumbing smoke: the option is
    // accepted and the local self-index path still works because
    // `announce_capabilities` indexes locally before sending.
    const a = await MeshNode.create({
      bindAddr: '127.0.0.1:0',
      psk: PSK,
      requireSignedCapabilities: true,
    });
    nodes.push(a);
    await a.announceCapabilities({ tags: ['local-only'] });
    expect(a.findPeers({ requireTags: ['local-only'] })).toContain(a.nodeId());
  });
});

File: net/crates/net/bindings/go/net/capabilities_test.go
// Tests for the capability announcement + filter surface (Stage G-2).
//
// Each node self-indexes its own announcement, so single-node
// round-trip covers the dict→core conversion plus filter predicate
// end-to-end. Multi-node propagation is covered by the Rust
// integration suite (`tests/capability_broadcast.rs`).

package net

import (
	"slices"
	"testing"
)

func newMeshForCaps(t *testing.T) *MeshNode {
	t.Helper()
	addr := reserveLocalUDPPort(t)
	m, err := NewMeshNode(MeshConfig{BindAddr: addr, PskHex: meshPsk})
	if err != nil {
		t.Fatalf("new mesh: %v", err)
	}
	return m
}

// ---------------------------------------------------------------------------
// Self-match round-trip
// ---------------------------------------------------------------------------

func TestAnnounce_ThenFind_SelfMatchesOnTag(t *testing.T) {
	m := newMeshForCaps(t)
	defer m.Shutdown()

	if err := m.AnnounceCapabilities(CapabilitySet{Tags: []string{"gpu", "prod"}}); err != nil {
		t.Fatalf("announce: %v", err)
	}
	peers, err := m.FindPeers(CapabilityFilter{RequireTags: []string{"gpu"}})
	if err != nil {
		t.Fatalf("find_peers: %v", err)
	}
	if !slices.Contains(peers, m.NodeID()) {
		t.Fatalf("own node id missing from find_peers result: %v", peers)
	}
}

func TestFindPeers_EmptyWhenFilterMismatches(t *testing.T) {
	m := newMeshForCaps(t)
	defer m.Shutdown()

	if err := m.AnnounceCapabilities(CapabilitySet{Tags: []string{"cpu"}}); err != nil {
		t.Fatalf("announce: %v", err)
	}
	peers, err := m.FindPeers(CapabilityFilter{RequireTags: []string{"gpu"}})
	if err != nil {
		t.Fatalf("find_peers: %v", err)
	}
	if len(peers) != 0 {
		t.Fatalf("expected empty peer list, got %v", peers)
	}
}

func TestFindPeers_WithoutAnnouncement_IsEmpty(t *testing.T) {
	m := newMeshForCaps(t)
	defer m.Shutdown()

	peers, err := m.FindPeers(CapabilityFilter{RequireTags: []string{"anything"}})
	if err != nil {
		t.Fatalf("find_peers: %v", err)
	}
	if len(peers) != 0 {
		t.Fatalf("expected empty peer list, got %v", peers)
	}
}

// ---------------------------------------------------------------------------
// Hardware / model / tool filters
// ---------------------------------------------------------------------------

func TestHardwareAndGpuFilter_Matches(t *testing.T) {
	m := newMeshForCaps(t)
	defer m.Shutdown()

	err := m.AnnounceCapabilities(CapabilitySet{
		Hardware: &HardwareCaps{
			CPUCores: 16,
			MemoryMB: 65536,
			GPU:      &GPUInfo{Vendor: "nvidia", Model: "h100", VRAMMB: 81920},
		},
		Tags: []string{"gpu"},
	})
	if err != nil {
		t.Fatalf("announce: %v", err)
	}

	peers, err := m.FindPeers(CapabilityFilter{
		RequireGPU:  true,
		GPUVendor:   "nvidia",
		MinVRAMMB:   40000,
		MinMemoryMB: 32768,
	})
	if err != nil {
		t.Fatalf("find_peers: %v", err)
	}
	if !slices.Contains(peers, m.NodeID()) {
		t.Fatalf("own id missing: %v", peers)
	}

	strict, err := m.FindPeers(CapabilityFilter{MinVRAMMB: 200_000})
	if err != nil {
		t.Fatalf("find_peers strict: %v", err)
	}
	if len(strict) != 0 {
		t.Fatalf("expected strict filter to miss, got %v", strict)
	}
}

func TestModelAndToolFilter_Matches(t *testing.T) {
	m := newMeshForCaps(t)
	defer m.Shutdown()

	err := m.AnnounceCapabilities(CapabilitySet{
		Models: []ModelCaps{{
			ModelID:       "llama-3.1-70b",
			Family:        "llama",
			ParametersBx10: 700,
			ContextLength: 128_000,
			Modalities:    []string{"text", "code"},
		}},
		Tools: []ToolCaps{{ToolID: "sql_exec", Name: "SQL Exec"}},
	})
	if err != nil {
		t.Fatalf("announce: %v", err)
	}

	peers, err := m.FindPeers(CapabilityFilter{RequireModels: []string{"llama-3.1-70b"}})
	if err != nil || !slices.Contains(peers, m.NodeID()) {
		t.Fatalf("model filter missed own node: err=%v peers=%v", err, peers)
	}
	peers, err = m.FindPeers(CapabilityFilter{RequireTools: []string{"sql_exec"}})
	if err != nil || !slices.Contains(peers, m.NodeID()) {
		t.Fatalf("tool filter missed own node: err=%v peers=%v", err, peers)
	}
	peers, err = m.FindPeers(CapabilityFilter{
		RequireModalities: []string{"code"},
		MinContextLength:  100_000,
	})
	if err != nil || !slices.Contains(peers, m.NodeID()) {
		t.Fatalf("modality filter missed own node: err=%v peers=%v", err, peers)
	}

	missing, err := m.FindPeers(CapabilityFilter{RequireModels: []string{"missing"}})
	if err != nil {
		t.Fatalf("find_peers missing-model query: %v", err)
	}
	if len(missing) != 0 {
		t.Fatalf("expected no match, got %v", missing)
	}
}

func TestEmptyAnnouncement_StillSelfIndexes(t *testing.T) {
	m := newMeshForCaps(t)
	defer m.Shutdown()

	if err := m.AnnounceCapabilities(CapabilitySet{}); err != nil {
		t.Fatalf("announce: %v", err)
	}
	peers, err := m.FindPeers(CapabilityFilter{})
	if err != nil {
		t.Fatalf("find_peers: %v", err)
	}
	if !slices.Contains(peers, m.NodeID()) {
		t.Fatalf("expected own node indexed, got %v", peers)
	}
}

// ---------------------------------------------------------------------------
// Vendor normalization helper
// ---------------------------------------------------------------------------

func TestNormalizeGPUVendor(t *testing.T) {
	cases := []struct {
		raw, want string
	}{
		{"NVIDIA", "nvidia"},
		{"Nvidia", "nvidia"},
		{"amd", "amd"},
		{"Apple", "apple"},
		{"qualcomm", "qualcomm"},
		{"intel", "intel"},
		{"bogus", "unknown"},
		{"", "unknown"},
	}
	for _, c := range cases {
		got, err := NormalizeGPUVendor(c.raw)
		if err != nil {
			t.Fatalf("normalize %q: %v", c.raw, err)
		}
		if got != c.want {
			t.Fatalf("normalize %q: want %q, got %q", c.raw, c.want, got)
		}
	}
}

// ---------------------------------------------------------------------------
// Constructor kwargs (capability_gc + signed)
// ---------------------------------------------------------------------------

func TestMeshConfig_AcceptsCapabilityKwargs(t *testing.T) {
	addr := reserveLocalUDPPort(t)
	m, err := NewMeshNode(MeshConfig{
		BindAddr:                  addr,
		PskHex:                    meshPsk,
		CapabilityGCIntervalMs:    120_000,
		RequireSignedCapabilities: true,
	})
	if err != nil {
		t.Fatalf("new mesh with cap kwargs: %v", err)
	}
	defer m.Shutdown()
	if m.NodeID() == 0 {
		t.Fatal("node id unset")
	}
}
